content/blog/2021/11/deep-dive-into-policy-logging.md

```diff
@@ -0,0 +1,221 @@
+between the guest (the policy) and the host (Kubewarden -- the `policy-server` or `kwctl` if we are
+running policies manually).
+
+Kubewarden has defined  a contract between policies (guests) and the host (Kubewarden) for
+performing [policy settings
+validation](https://docs.kubewarden.io/writing-policies/spec/02-settings.html), [policy
+validation](https://docs.kubewarden.io/writing-policies/spec/03-validating-policies.html), [policy
+mutation](https://docs.kubewarden.io/writing-policies/spec/04-mutating-policies.html) and logging.
+
+The waPC interface used for logging is a therefore a contract, because once you have built a policy,
+it should be possible to run it in future Kubewarden versions. In this sense, Kubewarden keeps this
+contract behind the SDK of your preferred language, so you don't have to deal with the details of
+how logging is implemented in Kubewarden. You just have to use your logging library of choice for
+the language you are working with.
+
+Let's look into how to take advantage of logging with Kubewarden in specific languages!
+
+## Policy Authors
+
+### Go
+
+We are going to use the [Go policy template](https://github.com/kubewarden/go-policy-template) as a
+starting point.
+
+Our Go SDK provides integration with the [`onelog`](https://github.com/francoispqt/onelog) library.
+When our policy is built for the WebAssembly target, it will send the logs to the host through waPC.
+Otherwise, it will just print them on stderr -- but this is only relevant if you happen to run your
+policy outside a Kubewarden runtime environment.
+
+One of the first things our policy does on its
+[`main.go`](https://github.com/kubewarden/go-policy-template/blob/a8dd2b374ce32ab01838d0c3a04efda1735599b3/main.go)
+file is to initialize the logger:
+
+```go
+var (
+	logWriter = kubewarden.KubewardenLogWriter{}
+	logger    = onelog.New(
+		&logWriter,
+		onelog.ALL, // shortcut for onelog.DEBUG|onelog.INFO|onelog.WARN|onelog.ERROR|onelog.FATAL
+	)
+)
+```
+
+We are then able to use `onelog` API in order to produce log messages. We could, for example,
+perform structured logging with debugging level:
+
+```go
+logger.DebugWithFields("validating object", func(e onelog.Entry) {
+    e.String("name", gjson.GetBytes(payload, "request.object.metadata.name").String())
+	e.String("namespace", gjson.GetBytes(payload, "request.object.metadata.namespace").String())
+})
+```
+
+Or, with info level:
+
+```go
+logger.InfoWithFields("validating object", func(e onelog.Entry) {
+    e.String("name", gjson.GetBytes(payload, "request.object.metadata.name").String())
+	e.String("namespace", gjson.GetBytes(payload, "request.object.metadata.namespace").String())
+})
+```
+
+What happens under the covers is that our Go SDK sends [every log event to the `kubewarden` host
+through
+waPC](https://github.com/kubewarden/policy-sdk-go/blob/95c4b93adf6830862b85f953bc09d31de7490872/log_writer_wasi.go#L15).
```

Comment:
I have mixed feelings about these sections. I think we can drop them for the sake of reducing the word count on the whole article (and keep the reader focused and engaged)

Reply:
I would prefer to keep them, since if you are interested in a language it's one click away, it's not like we are embedding more code making the post huge.

Also, rendered does not seem so long to me.

Reply:
I also think it adds value because if you are interested as a reader you can see that basically they approach waPC in the same way.

Reply:
I'm happy with them, helped me learn the specifics.

Reply:
Fine with me :)